components/script/dom/htmldetailselement.rs

```diff
@@ -60,6 +90,132 @@ impl HTMLDetailsElement {
     pub(crate) fn toggle(&self) {
         self.SetOpen(!self.Open());
     }
+
+    fn shadow_tree(&self, can_gc: CanGc) -> Ref<'_, ShadowTree> {
+        if !self.upcast::<Element>().is_shadow_host() {
+            self.create_shadow_tree(can_gc);
+        }
+
+        Ref::filter_map(self.shadow_tree.borrow(), Option::as_ref)
+            .ok()
+            .expect("UA shadow tree was not created")
+    }
+
+    fn create_shadow_tree(&self, can_gc: CanGc) {
+        let document = self.owner_document();
+        let root = self
+            .upcast::<Element>()
+            .attach_shadow(
+                IsUserAgentWidget::Yes,
+                ShadowRootMode::Closed,
+                false,
+                SlotAssignmentMode::Manual,
+            )
+            .expect("Attaching UA shadow root failed");
+
+        let summary = HTMLSlotElement::new(local_name!("slot"), None, &document, None, can_gc);
+        root.upcast::<Node>()
+            .AppendChild(summary.upcast::<Node>())
+            .unwrap();
+
+        let fallback_summary =
+            HTMLElement::new(local_name!("summary"), None, &document, None, can_gc);
+        fallback_summary
+            .upcast::<Node>()
+            .SetTextContent(Some(DEFAULT_SUMMARY.into()), can_gc);
+        summary
+            .upcast::<Node>()
+            .AppendChild(fallback_summary.upcast::<Node>())
+            .unwrap();
+
+        let descendants = HTMLSlotElement::new(local_name!("slot"), None, &document, None, can_gc);
+        root.upcast::<Node>()
+            .AppendChild(descendants.upcast::<Node>())
+            .unwrap();
+
+        let _ = self.shadow_tree.borrow_mut().insert(ShadowTree {
+            summary: summary.as_traced(),
+            descendants: descendants.as_traced(),
+            implicit_summary: fallback_summary.as_traced(),
+        });
+        self.upcast::<Node>()
+            .dirty(crate::dom::node::NodeDamage::OtherNodeDamage);
+    }
+
+    pub(crate) fn find_corresponding_summary_element(&self) -> Option<DomRoot<HTMLElement>> {
+        for node in self.upcast::<Node>().children() {
+            if let Some(html_element) = node.downcast::<HTMLElement>() {
+                if html_element.upcast::<Element>().local_name() == &local_name!("summary") {
+                    return Some(DomRoot::from_ref(html_element));
+                }
+            }
+        }
+        None
```

Comment:
Could this use [`find()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find) perhaps?

Reply:
While trying to explain why it isn't possible i realized that it actually is... Will fix.